pkg/datapath/iptables/reconciler_test.go

```diff
@@ -0,0 +1,241 @@
+	store.Update(func(n *node.LocalNode) {
+		n.IPAddresses = []types.Address{{IP: netip.MustParseAddr("1.1.1.1").AsSlice()}}
+	})
+
+	txn := db.WriteTxn(devices)
+	if _, _, err := devices.Insert(txn, &tables.Device{
+		Index:    1,
+		Name:     "test-1",
+		Selected: true,
+	}); err != nil {
+		t.Fatal(err)
+	}
+	txn.Commit()
+
+	// start the reconciliation loop
+	go func() {
+		defer close(errs)
+		errs <- reconciliationLoop(ctx, health, true, store, db, devices, updateFunc)
+	}()
+
+	// wait for reconciler to update the state as desired
+	var lastErr error
+	if err := testutils.WaitUntil(func() bool {
+		expected := desiredState{
+			firstInit:    true,
+			installRules: true,
+			devices:      []*tables.Device{{Name: "test-1"}},
+			localNode: node.LocalNode{Node: types.Node{
+				IPAddresses: []types.Address{{IP: netip.MustParseAddr("1.1.1.1").AsSlice()}},
+			}},
+		}
+
+		curState, ok := <-states
+		if !ok {
+			lastErr = fmt.Errorf("state updates channel closed unexpectedly")
+			return false
+		}
+		if err := matchState(*curState, expected); err != nil {
+			lastErr = err
+			return false
+		}
+		return true
+	}, 10*time.Second); err != nil {
+		t.Fatal(lastErr)
+	}
+
+	// test devices update
+	txn = db.WriteTxn(devices)
+	devices.Insert(txn, &tables.Device{
+		Index:    2,
+		Name:     "test-2",
+		Selected: true,
+	})
+	txn.Commit()
+
+	if err := testutils.WaitUntil(func() bool {
+		expected := desiredState{
+			firstInit:    false,
+			installRules: true,
+			devices:      []*tables.Device{{Name: "test-1"}, {Name: "test-2"}},
+			localNode: node.LocalNode{Node: types.Node{
+				IPAddresses: []types.Address{{IP: netip.MustParseAddr("1.1.1.1").AsSlice()}},
+			}},
+		}
+
+		curState, ok := <-states
+		if !ok {
+			lastErr = fmt.Errorf("state updates channel closed unexpectedly")
+			return false
+		}
+		if err := matchState(*curState, expected); err != nil {
+			lastErr = err
+			return false
+		}
+		return true
+	}, 10*time.Second); err != nil {
+		t.Fatal(lastErr)
+	}
+
+	// test local node update
+	store.Update(func(n *node.LocalNode) {
+		n.IPAddresses = []types.Address{{IP: netip.MustParseAddr("2.2.2.2").AsSlice()}}
+	})
+
+	if err := testutils.WaitUntil(func() bool {
+		expected := desiredState{
+			firstInit:    false,
+			installRules: true,
+			devices:      []*tables.Device{{Name: "test-1"}, {Name: "test-2"}},
+			localNode: node.LocalNode{Node: types.Node{
+				IPAddresses: []types.Address{{IP: netip.MustParseAddr("2.2.2.2").AsSlice()}},
+			}},
+		}
+
+		curState, ok := <-states
+		if !ok {
+			lastErr = fmt.Errorf("state updates channel closed unexpectedly")
+			return false
+		}
+		if err := matchState(*curState, expected); err != nil {
+			lastErr = err
+			return false
+		}
+		return true
+	}, 10*time.Second); err != nil {
+		t.Fatal(lastErr)
+	}
+
+	cancel()
+
+	if err := <-errs; err != nil {
+		t.Fatal(err)
+	}
+
+	close(states)
+
+	// drain the states channel
+	for range states {
```

Comment:
Drain after `close` is a no-op. Also is it safe to `close` above? Do we really know nothing will send to it?

Reply:
This is a leftover from a previous version, I'll remove it. :+1: